Generate offline map using android jetpack's workmanager

kaushikrw opened the pull request:
## Description
PR to add a new Kotlin sample _"Generate offline map using workmanager"_ in `Edit and Manage Data` category.

## Links and Data
Sample Epic: `runtime/kotlin/issues/1731`

## What To Review
-  Review the code to make sure it is easy to follow like other samples on Android
- `README.md` and `README.metadata.json` files

## How to Test
Run the sample on the sample viewer or the repo.
Please check the README for more information - [README.md](https://github.com/Esri/arcgis-maps-sdk-kotlin-samples/tree/kaushikrw/generate-offline-map-workmanager/generate-offline-map-using-workmanager)

## Notes
If the checker fails on the README stage at `raise Exception(f'Wrong letter case for word: "{word}" in title.')` its because the checker expects only the first word of the title to be upper case. Please ignore it as we want to preserve the title casing for `Android Jetpack's WorkManager`

<!-- OPTIONAL
## Screenshots
-->

shubham7109 commented on generate-offline-map-using-workmanager/README.md:
```diff
@@ -0,0 +1,68 @@
+
+## Tags
+
+download, offline, save, web map
```
Would be nice to add: "workmanager, service, background, notification" to the tags and the meta-data.

shubham7109 commented on generate-offline-map-using-workmanager/src/main/java/com/esri/arcgismaps/sample/generateofflinemapusingworkmanager/MainActivity.kt:
```diff
@@ -0,0 +1,441 @@
+
+    // used to uniquely identify the work request so that only one worker is active at a time
+    // also allows us to query and observe work progress
+    private val uniqueWorkName = "offlineJobWorker"
```
```suggestion
    private val uniqueWorkName = "ArcgisMaps.Sample.OfflineMapJob.Worker"
```

shubham7109 commented on generate-offline-map-using-workmanager/src/main/java/com/esri/arcgismaps/sample/generateofflinemapusingworkmanager/MainActivity.kt:
```diff
@@ -0,0 +1,441 @@
+        lifecycleScope.launch {
+            // collect the live data flow to get the latest work info list
+            liveDataFlow.collect { workInfoList ->
+                if (workInfoList.size > 0) {
```
```suggestion
                if (workInfoList.isNotEmpty()) {
```

shubham7109 commented on generate-offline-map-using-workmanager/src/main/java/com/esri/arcgismaps/sample/generateofflinemapusingworkmanager/MainActivity.kt:
```diff
@@ -0,0 +1,441 @@
+                workManager.pruneWork()
+                // display the offline map loaded message
+                showMessage("Loaded offline map. Map saved at: $offlineMapPath")
+            }
```
else, "Offline map does not exist at path: "

shubham7109 commented on generate-offline-map-using-workmanager/src/main/java/com/esri/arcgismaps/sample/generateofflinemapusingworkmanager/WorkerNotification.kt:
```diff
@@ -0,0 +1,143 @@
+            // sets the onclick action to launch the mainActivityIntent
+            .setContentIntent(mainActivityIntent)
+            // sets it to show the notification immediately
+            .setForegroundServiceBehavior(NotificationCompat.FOREGROUND_SERVICE_IMMEDIATE)
```
```suggestion
            .setForegroundServiceBehavior(NotificationCompat.FOREGROUND_SERVICE_IMMEDIATE)
            .addAction(R.drawable.ic_launcher_foreground,"Cancel",cancelIntent)
```
Suggestion, but would be okay without it as well.

Since the sample showcases the job worker and running the offline map job in the background, it would kinda be nice to show the cancellation of the job also handled in the background. There is a helpful way to add "[actions](https://developer.android.com/develop/ui/views/notifications/build-notification#Actions)" to notifications which takes also would use a `PendingIntent`.

kaushikrw replied:
Thanks for the suggestion. I've added a cancellation action.

shubham7109 commented on generate-offline-map-using-workmanager/src/main/res/values/strings.xml:
```diff
@@ -0,0 +1,11 @@
+    <string name="offlineJobJsonFile">/offlineJobJson</string>
+    <string name="notification_channel_name">Offline Map Job Notifications</string>
+    <string name="notification_channel_description">Shows notifications for offline map job progress</string>
+    <string name="notification_title">Offline Map Download</string>
```
```suggestion
    <string name="notification_title>ArcGIS Maps Sample: Offline Map Download</string>
```

shubham7109 reviewed (changes requested):
Really nice sample here @kaushikrw. Took me a bit to read up doc on how the `WorkManager` works (pun indented). Also, I'm sure users will appreciate the separation in code for the sample, notification, and offline job worker. Left couple comments

gunt0001 commented on generate-offline-map-using-workmanager/src/main/java/com/esri/arcgismaps/sample/generateofflinemapusingworkmanager/MainActivity.kt:
```diff
@@ -0,0 +1,441 @@
+        // create the json file
+        val offlineJobJsonFile = File(offlineJobJsonPath)
+        // serialize the offlineMapJob into the file
+        offlineJobJsonFile.writeText(offlineMapJob.toJson())
```
Is there a need to write the serialized job to a file? You could just pass the JSON string directly to the work request.

kaushikrw replied:
There is a [MAX_DATA_BYTES](https://developer.android.com/reference/kotlin/androidx/work/Data#MAX_DATA_BYTES()) size limit that is enforced on the payloads for `Data` that we add to a `WorkRequest`. It's about ~10Kb and often the JSON string exceeds that.

This is added to the function level comments too - https://github.com/Esri/arcgis-maps-sdk-kotlin-samples/blob/5911bbe08a428ed06136864d5b054092165bc8c6/generate-offline-map-using-workmanager/src/main/java/com/esri/arcgismaps/sample/generateofflinemapusingworkmanager/MainActivity.kt#L229-L234

gunt0001 commented on generate-offline-map-using-workmanager/src/main/java/com/esri/arcgismaps/sample/generateofflinemapusingworkmanager/OfflineJobWorker.kt:
```diff
@@ -0,0 +1,153 @@
+
+    override suspend fun doWork(): Result {
+        // CoroutineScope to run the GenerateOfflineMapJob's state flows on
+        val coroutineScope = CoroutineScope(Dispatchers.Default)
```
I think you are better off using [coroutineScope](https://kotlinlang.org/api/kotlinx.coroutines/kotlinx-coroutines-core/kotlinx.coroutines/coroutine-scope.html) function here

gunt0001 commented on generate-offline-map-using-workmanager/src/main/java/com/esri/arcgismaps/sample/generateofflinemapusingworkmanager/OfflineJobWorker.kt:
```diff
@@ -0,0 +1,153 @@
+            }
+
+            // wait for the completion of the deferred value and return its result
+            deferred.await()
```
No need to use a CompletableDeferred here. You have two options:
1. Use `Job.result()` here.
2. Rely on the use of the `coroutineScope` function as suggested above. `coroutineScope` will not resume until all of its child coroutines have completed. So within the `coroutineScope` block you could simply collect progress and status event and as soon as you get a job status of `Succeeded` or `Failed`, you terminate the collection of status and progress flows. You can do this by using predicates while collecting, for example [takeWhile](https://kotlinlang.org/api/kotlinx.coroutines/kotlinx-coroutines-core/kotlinx.coroutines.flow/take-while.html).

kaushikrw replied:
Thanks for the feedback! I've refactored it to use `Job.result()` and `takeWhile` and removed the use of a `CompletableDeferred`

gunt0001 commented on generate-offline-map-using-workmanager/src/main/java/com/esri/arcgismaps/sample/generateofflinemapusingworkmanager/OfflineJobWorker.kt:
```diff
@@ -0,0 +1,147 @@
+            // can be run on the default Dispatchers.Default context
+            generateOfflineMapJob.start()
+
+            // launch a progress collector in a new CoroutineScope
+            progressCollectorJob = CoroutineScope(Dispatchers.Default).launch {
+                // collect on progress until the job has completed in a success/failure
+                generateOfflineMapJob.progress.takeWhile {
+                    generateOfflineMapJob.status.value != JobStatus.Failed
+                        && generateOfflineMapJob.status.value != JobStatus.Succeeded
+                }.collect { progress ->
+                    // update the worker progress
+                    setProgress(workDataOf("Progress" to progress))
+                    // update the ongoing progress notification
+                    setForeground(createForegroundInfo(progress))
+                }
+            }
+
+            // suspends until the generateOfflineMapJob has completed
+            val result = generateOfflineMapJob.result()
```
using one of the coroutine scope functions to create a new scope is better than creating a new CoroutineScope instance directly here, since the use of a coroutine scope function ensures that the coroutine context is inherited from the parent, thus preserving structured concurrency. It also means you won't have to manage the `progressCollectorJob` with regards to cancellation.  I suggest the following:
```suggestion
            val result = coroutineScope {
                // launch progress collection in a new coroutine, for this we need a new coroutine scope, created with the coroutineScope() function
                launch {
                     // collect progress until the job has completed with success/failure
                     generateOfflineMapJob.progress.takeWhile {
                     generateOfflineMapJob.status.value != JobStatus.Failed
                        && generateOfflineMapJob.status.value != JobStatus.Succeeded
                     }.collect { progress ->
                      // update the worker progress
                      setProgress(workDataOf("Progress" to progress))
                      // update the ongoing progress notification
                      setForeground(createForegroundInfo(progress))
                  }
                }
                
                // suspends until the generateOfflineMapJob has completed
                generateOfflineMapJob.result()
            }
```

gunt0001 commented on generate-offline-map-using-workmanager/src/main/java/com/esri/arcgismaps/sample/generateofflinemapusingworkmanager/OfflineJobWorker.kt:
```diff
@@ -0,0 +1,147 @@
+            Result.failure()
+        } finally {
+            // cancel the progressCollectorJob if it is running
+            progressCollectorJob?.cancel()
```
No need for this if you use `coroutineScope()` function

kaushikrw commented on generate-offline-map-using-workmanager/README.md:
```diff
@@ -0,0 +1,66 @@
+# Generate offline map using workmanager
```
@shubham7109 Can the title and the project name be different? Or would you suggest changing the project/directory name too?

shubham7109 replied:
I would suggest changing the module project and directory names as well, since there are scripts that require the names to be consistent. I realize this would require many changes but it would be necessary.

kaushikrw replied:
The `Jetpack's` apostrophe is causing issues in multiple places. Shall I omit that?

shubham7109 replied:
Sure feel free to omit it. Also, I think it would be a good idea updating the description:
>Take a web map offline as a persistent background task using Android Jetpack's [WorkManager](https://developer.android.com/topic/libraries/architecture/workmanager)

shubham7109 commented on generate-offline-map-using-android-jetpack-workmanager/README.md:
```diff
@@ -0,0 +1,66 @@
+
+Take a web map offline as a persistent background task using Android Jetpack's [WorkManager](https://developer.android.com/topic/libraries/architecture/workmanager)
+
+![Image of generate offline map](generate-offline-map.png)
```
```suggestion
![Image of generate offline map using Android Jetpack WorkManager](generate-offline-map-using-android-jetpack-workmanager.png)
```

shubham7109 commented on generate-offline-map-using-android-jetpack-workmanager/README.metadata.json:
```diff
@@ -0,0 +1,40 @@
+    "formal_name": "GenerateOfflineMapUsingAndroidJetpackWorkmanager",
+    "ignore": false,
+    "images": [
+        "generate-offline-map.png"
```
```suggestion
        "generate-offline-map-using-android-jetpack-workmanager.png"
```

kaushikrw commented:
Thanks @shubham7109 and @gunt0001 for the feedback. The PR is ready for a re-review now.

The renaming caused some issues with the sample viewer. I have fixed that now and tested it. The check might be failing but I think that is okay, I've added the reason for this in the PR description.

gunt0001 commented on generate-offline-map-using-android-jetpack-workmanager/src/main/java/com/esri/arcgismaps/sample/generateofflinemapusingandroidjetpackworkmanager/MainActivity.kt:
```diff
@@ -0,0 +1,443 @@
+
+            // limit the map scale to the largest layer scale
+            map.maxScale = map.operationalLayers[6].maxScale ?: 0.0
+            map.minScale = map.operationalLayers[6].minScale ?: 0.0
```
Will the hard coded index of `6` always apply? It might be saver to find the specific layer, based on a name or similar...

kaushikrw replied:
Fetching a layer based on the name makes sense, changed it to use that.

gunt0001 commented on generate-offline-map-using-android-jetpack-workmanager/src/main/java/com/esri/arcgismaps/sample/generateofflinemapusingandroidjetpackworkmanager/OfflineJobWorker.kt:
```diff
@@ -0,0 +1,147 @@
+            // capture and log if exception occurs
+            Log.e(TAG, "Offline map job failed:", exception)
+            // a CancellationException is raised if the work is cancelled manually by the user
+            if (exception !is CancellationException) {
```
If it's a `CancellationException`, you need to rethrow it, otherwise you break structured concurrency.

kaushikrw replied:
In this case, since the `CoroutineWorker.doWork` is the parent/root is re-throwing necessary? I saw [doc](https://developer.android.com/guide/background/persistent/threading/coroutineworker) says as follows which I interpreted as not having to rethrow?  - 

`CoroutineWorker handles stoppages automatically by cancelling the coroutine and propagating the cancellation signals. You don't need to do anything special to handle [work stoppages](https://developer.android.com/topic/libraries/architecture/workmanager/how-to/managing-work#cancelling).`

gunt0001 replied:
@kaushikrw - I think this is different from what that doc is talking about. If you swallow the CancellationException, code that may run after this suspend function would be executed, even if the coroutine was cancelled. This would break the behaviour of structured concurrency. A longer discussion of this issue [here](https://github.com/Kotlin/kotlinx.coroutines/issues/1814), related to the use of `runCatching`.

kaushikrw replied:
I think that make sense, since I'm not doing anything special to handle a`CancellationException` and the WorkManager ignores the returned result in case of cancellation. Modified it to rethrow if it's a `CancellationException`.

But the WorkManager [samples](https://github.com/googlecodelabs/android-workmanager/blob/advanced/app/src/main/java/com/example/background/workers/BlurWorker.kt) from Google all seem to be actually swallowing CancellationExceptions. I wonder if it's due to an inherent characteristic of WorkManager cancellation mechanism or an oversight in their samples.

gunt0001 replied:
Interesting. I would assume it's an oversight in their samples. It was also metioned in the [linked issue](https://github.com/Kotlin/kotlinx.coroutines/issues/1814) that many samples ignore the fact that CancellationException should not be swallowed. It's possible that it is even ok in the case of `CoroutineWorker.doWork`, as long as the calling code does not run any other suspend functions on the coroutine that calls `doWork`. But from my POV that would be an assumption we should not make in our code.

gunt0001 reviewed (approved):
@kaushikrw - other than the two remaining comments, this looks good to me.

shubham7109 reviewed (approved):
Looks good, runs well with the sample viewer 👍

gunt0001 commented on generate-offline-map-using-android-jetpack-workmanager/src/main/java/com/esri/arcgismaps/sample/generateofflinemapusingandroidjetpackworkmanager/MainActivity.kt:
```diff
@@ -0,0 +1,450 @@
+            // enable the take map offline button only after the map is loaded
+            takeMapOfflineButton.isEnabled = true
+
+            // get the Control Valve layer, which is the largest layer in the
```
Would a better wording be?
```suggestion
            // get the Control Valve layer, which is the top most layer in the
```

shubham7109 replied:
Not sure if "top most" would be accurate. Seems to be the "Control Valve" layer would be 4th layer on the [web-map](https://arcgisruntime.maps.arcgis.com/home/item.html?id=acc027394bc84c2fb04d1ed317aac674). We can just drop the sentence: 
```kotlin
// get the Control Valve layer
```

gunt0001 commented on generate-offline-map-using-android-jetpack-workmanager/src/main/java/com/esri/arcgismaps/sample/generateofflinemapusingandroidjetpackworkmanager/MainActivity.kt:
```diff
@@ -0,0 +1,450 @@
+                    layer.name == "Control Valve"
+                } ?: return@launch showMessage("Error finding Control Valve layer")
+
+            // limit the map scale to the largest layer scale
```
```suggestion
            // limit the map scale to the top most layer's scale
```

shubham7109 replied:
Suggestion: 
```
// limit the map scale to the layer's scale
```

gunt0001 reviewed (approved):
@kaushikrw - looks great, some final very minor comments.

kaushikrw commented:
Thanks @gunt0001 and @shubham7109. Merging the PR now.
